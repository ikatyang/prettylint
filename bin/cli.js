#!/usr/bin/env node

import { run } from "../lib/cli.js";

await run(process.argv.slice(2));
