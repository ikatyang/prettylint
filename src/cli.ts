import * as fs from "node:fs";
import * as path from "node:path";
import { globby } from "globby";
import ignore from "ignore";
import meow from "meow";
import { lint, LintResult } from "./lint.js";

type Formatter = (results: LintResult[]) => string;

export const helpMessage: string = `
  Usage
    $ prettylint <glob> ...

  Options
    --no-config           Do not load config file.
    --config <path>       Specify the config file.
    --fix                 Fix linting errors.
    --format <path>       Specify the module to format output.
                          Defaults to "eslint-formatter-pretty".
    --ignore-path <path>  Specify the .ignore file.
                          Defaults to ".prettierignore".
    --silent              Do not print message.
    -h, --help            Show help.
    -v, --version         Show version.

  Examples
    $ prettylint "src/**/*.{js,ts}"
    $ prettylint "src/**/*.{js,ts}" --fix
    $ prettylint "src/**/*.{js,ts}" --ignore-path .gitignore
`;

export async function run(argv: string[]) {
  const { input, flags } = meow(helpMessage, {
    importMeta: import.meta,
    argv,
    flags: {
      fix: {
        type: "boolean",
        default: false,
      },
      silent: {
        type: "boolean",
        default: false,
      },
      config: {
        type: "string",
      },
      ignorePath: {
        type: "string",
      },
      format: {
        type: "string",
        default: "eslint-formatter-pretty",
      },
      help: {
        aliases: ["h"],
      },
      version: {
        aliases: ["v"],
      },
    },
  });

  const ignorer = ignore.default();
  const defaultIgnorePath = ".prettierignore";

  if (flags.ignorePath === undefined) {
    if (!fs.existsSync(defaultIgnorePath)) {
      input.push("!**/node_modules/**");
    } else {
      ignorer.add(await readFile(defaultIgnorePath));
    }
  } else {
    ignorer.add(await readFile(flags.ignorePath));
  }

  const filenames = ignorer.filter(
    (await globby(input, { dot: true })) //
      .map((filename) => path.relative(".", filename)),
  );

  let errorCount = 0;
  let warningCount = 0;

  const results = await Promise.all(
    filenames.map(async (filename) => {
      const content = await readFile(filename);
      const result = await lint(filename, content, { config: flags.config });

      errorCount += result.errorCount;
      warningCount += result.warningCount;

      if (result.output === undefined || !flags.fix) {
        return result;
      }

      await writeFile(filename, result.output);

      const { output, ...newResult } = result;
      return { ...newResult, messages: [], warningCount: 0 };
    }),
  );

  if (!flags.silent) {
    const formatter: Formatter = (await import(flags.format)).default;
    const message = formatter(results);
    if (message.length !== 0) {
      console.log(message);
    }
  }

  if (errorCount !== 0 || (!flags.fix && warningCount !== 0)) {
    process.exitCode = 1;
  }

  return results;
}

async function readFile(filename: string) {
  return new Promise<string>((resolve, reject) => {
    fs.readFile(filename, { encoding: "utf8" }, (error, data) => {
      /* c8 ignore start */
      if (error) {
        reject(error);
      } /* c8 ignore stop */ else {
        resolve(data);
      }
    });
  });
}

async function writeFile(filename: string, content: string) {
  return new Promise<void>((resolve, reject) => {
    fs.writeFile(filename, content, (error) => {
      /* c8 ignore start */
      if (error) {
        reject(error);
      } /* c8 ignore stop */ else {
        resolve();
      }
    });
  });
}
