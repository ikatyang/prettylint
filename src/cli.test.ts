import * as fs from 'fs';
import * as path from 'path';
import strip_ansi = require('strip-ansi');
import { run } from './cli';

const base_dir = path.resolve(__dirname, '../fixtures/cli');

let status: number | undefined;
let stdout: string | undefined;
let writes: { filename: string; content: string }[];

setup();

test('ignorePath: default, .prettierignore exists', async () =>
  run_cli(['**/*.js'], {
    cwd: `${base_dir}/ignore-path/default-with-prettierignore`,
  }).then(results => {
    expect(status).toBe(0);
    expect(results).toMatchSnapshot();
  }));

test('ignorePath: default, .prettierignore does not exist', async () =>
  run_cli(['**/*.js'], {
    cwd: `${base_dir}/ignore-path/default-without-prettierignore`,
  }).then(results => {
    expect(status).toBe(0);
    expect(results).toMatchSnapshot();
  }));

test('ignorePath: ".ignore"', async () =>
  run_cli(['**/*.js', '--ignore-path', '.ignore'], {
    cwd: `${base_dir}/ignore-path/custom-ignore`,
  }).then(results => {
    expect(status).toBe(0);
    expect(results).toMatchSnapshot();
  }));

test('fix: true', async () =>
  run_cli(['**/*.js', '--fix'], { cwd: `${base_dir}/fix` }).then(() => {
    expect(status).toBe(0);
    expect(stdout).toBe(undefined);
    expect(writes).toMatchSnapshot();
  }));

test('format: default', async () =>
  run_cli(['**/*.js'], { cwd: `${base_dir}/format/default` }).then(() => {
    expect(status).toBe(1);
    expect(stdout).toMatchSnapshot();
  }));

test('silent: true', async () =>
  run_cli(['**/*.js', '--silent'], { cwd: `${base_dir}/silent` }).then(() => {
    expect(status).toBe(1);
    expect(stdout).toBe(undefined);
  }));

interface RunCliOptions {
  cwd: string;
}

async function run_cli(argv: string[], options: RunCliOptions) {
  process.chdir(options.cwd);
  return run(argv).then(results => {
    status =
      status !== undefined
        ? status
        : // tslint:disable-next-line:strict-type-predicates
          process.exitCode !== undefined ? process.exitCode : 0;
    stdout = stdout === undefined ? undefined : strip_ansi(stdout);
    return results;
  });
}

function setup() {
  const original_cwd = process.cwd();
  const original_env_ci = process.env.CI;
  const original_exit_code = process.exitCode;

  beforeAll(() => {
    jest.spyOn(console, 'log').mockImplementation((...messages: string[]) => {
      if (!is_exited()) {
        stdout = `${stdout === undefined ? '' : stdout}${messages.join(' ')}\n`;
      }
    });
    jest
      .spyOn(fs, 'writeFile')
      .mockImplementation(
        (filename: string, content: string, callback: () => void) => {
          if (!is_exited()) {
            writes.push({ filename, content });
          }
          callback();
        },
      );
    jest.spyOn(process, 'exit').mockImplementation((exit_code?: number) => {
      if (!is_exited()) {
        status = exit_code === undefined ? 0 : exit_code;
      }
    });

    function is_exited() {
      // tslint:disable-next-line:strict-type-predicates
      return status !== undefined || process.exitCode !== undefined;
    }
  });

  beforeEach(() => {
    process.env.CI = 'true';
    process.exitCode = undefined as any;

    status = undefined;
    stdout = undefined;
    writes = [];
  });

  afterAll(() => {
    process.chdir(original_cwd);
    process.exitCode = original_exit_code;

    if (original_env_ci !== undefined) {
      process.env.CI = original_env_ci;
    } else {
      delete process.env.CI;
    }
  });
}
