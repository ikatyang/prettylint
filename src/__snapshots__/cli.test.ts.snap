// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fix: true 1`] = `
[
  {
    "content": "const x = 123;
",
    "filename": "unformatted.js",
  },
]
`;

exports[`format: default 1`] = `
"
  unformatted.js:0:7
  ⚠  0:7  Replace = with ·=·      prettier/prettier

  syntax-error.js:1:7
  ✖  1:7  Unexpected token (1:7)  prettier/prettier

  1 warning
  1 error

"
`;

exports[`ignorePath: ".ignore" 1`] = `
[
  {
    "errorCount": 0,
    "filePath": "non-ignored.js",
    "messages": [],
    "source": "",
    "warningCount": 0,
  },
  {
    "errorCount": 0,
    "filePath": "node_modules/non-ignored.js",
    "messages": [],
    "source": "",
    "warningCount": 0,
  },
]
`;

exports[`ignorePath: default, .prettierignore does not exist 1`] = `
[
  {
    "errorCount": 0,
    "filePath": "non-ignored.js",
    "messages": [],
    "source": "",
    "warningCount": 0,
  },
  {
    "errorCount": 0,
    "filePath": "folder/non-ignored.js",
    "messages": [],
    "source": "",
    "warningCount": 0,
  },
]
`;

exports[`ignorePath: default, .prettierignore exists 1`] = `
[
  {
    "errorCount": 0,
    "filePath": "non-ignored.js",
    "messages": [],
    "source": "",
    "warningCount": 0,
  },
  {
    "errorCount": 0,
    "filePath": "node_modules/non-ignored.js",
    "messages": [],
    "source": "",
    "warningCount": 0,
  },
]
`;
